src/dex/uniswap-v3/uniswap-v3.ts

```diff
@@ -0,0 +1,824 @@
+    if (!_pairs) {
+      this.logger.warn(
+        `did not find poolconfig in for key ${this.dexmapKey} ${poolKey}`,
+      );
+      return;
+    }
+
+    const poolInfo: PoolPairsInfo = JSON.parse(_pairs);
+
+    this.getPool(
+      poolInfo.token0,
+      poolInfo.token1,
+      BigInt(poolInfo.fee),
+      blockNumber,
+    );
+  }
+
+  async getPoolIdentifiers(
+    srcToken: Token,
+    destToken: Token,
+    side: SwapSide,
+    blockNumber: number,
+  ): Promise<string[]> {
+    const _srcToken = this.dexHelper.config.wrapETH(srcToken);
+    const _destToken = this.dexHelper.config.wrapETH(destToken);
+
+    const [_srcAddress, _destAddress] = this._getLoweredAddresses(
+      _srcToken,
+      _destToken,
+    );
+
+    if (_srcAddress === _destAddress) return [];
+
+    const pools = (
+      await Promise.all(
+        this.supportedFees.map(async fee =>
+          this.getPool(_srcAddress, _destAddress, fee, blockNumber),
+        ),
+      )
+    ).filter(pool => pool);
+
+    if (pools.length === 0) return [];
+
+    return pools.map(pool =>
+      this.getPoolIdentifier(_srcAddress, _destAddress, pool!.feeCode),
+    );
+  }
+
+  async getStateFromRpc(
+    from: Token,
+    to: Token,
+    amounts: bigint[],
+    side: SwapSide,
+    pools: UniswapV3EventPool[],
+  ): Promise<ExchangePrices<UniswapV3Data> | null> {
+    if (pools.length === 0) {
+      return null;
+    }
+    pools.forEach(pool => {
+      this.logger.warn(
+        `[${this.network}][${pool.parentName}] fallback to rpc for ${pool.name}`,
+      );
+    });
+
+    const unitVolume = getBigIntPow(
+      (side === SwapSide.SELL ? from : to).decimals,
+    );
+
+    const chunks = amounts.length - 1;
+
+    const _width = Math.floor(chunks / this.config.chuncksCount);
+
+    const _amounts = [unitVolume].concat(
+      Array.from(Array(this.config.chuncksCount).keys()).map(
+        i => amounts[(i + 1) * _width],
+      ),
+    );
+
+    const calldata = pools.map(pool =>
+      _amounts.map(_amount => ({
+        target: this.config.quoter,
+        gasLimit: UNISWAPV3_QUOTE_GASLIMIT,
+        callData:
+          side === SwapSide.SELL
+            ? this.quoterIface.encodeFunctionData('quoteExactInputSingle', [
+                from.address,
+                to.address,
+                pool.feeCodeAsString,
+                _amount.toString(),
+                0, //sqrtPriceLimitX96
+              ])
+            : this.quoterIface.encodeFunctionData('quoteExactOutputSingle', [
+                from.address,
+                to.address,
+                pool.feeCodeAsString,
+                _amount.toString(),
+                0, //sqrtPriceLimitX96
+              ]),
+      })),
+    );
+
+    const data = await this.uniswapMulti.methods
+      .multicall(calldata.flat())
+      .call();
+
+    const decode = (j: number): bigint => {
+      if (!data.returnData[j].success) return 0n;
```

Comment:
Maybe we should log this as well? Because I don't think we expect many calls to fail. But if it is failing, something is wrong.